base/binaryplatforms.jl

```diff
@@ -7,6 +7,62 @@ export AbstractPlatform, Platform, HostPlatform, platform_dlext, tags, arch, os,
        select_platform, platforms_match, platform_name
 import .Libc.Libdl
 
+### Submodule with information about CPU features
+module CPUID
+
+export cpu_isa
+
+"""
+    ISA(features::Set{UInt32})
+
+A structure which represents the Instruction Set Architecture (ISA) of a
+computer.  It holds the `Set` of features of the CPU.
+
+The numerical values of the features are automatically generated from the C
+source code of Julia and stored in the `features_h.jl` Julia file.
+"""
+struct ISA
+    features::Set{UInt32}
+end
+
+Base.:<=(a::ISA, b::ISA) = a.features <= b.features
+Base.:<(a::ISA,  b::ISA) = a.features <  b.features
+Base.isless(a::ISA,  b::ISA) = a < b
+
+include("features_h.jl")
+
+# Sets of features for the x86-64 microarchitectures come from
+# https://gcc.gnu.org/onlinedocs/gcc/x86-Options.html.  Keep in sync with
+# `arch_march_isa_mapping`.
+const ISAs_by_family = Dict(
+    "x86_64" => (
```

Comment:
Does it make sense to make these enums so you know from the type what possible the possible values are?

Reply:
I like Dict's better so that we can add to them via Pkg hooks in the future.

Reply:
I didn't mean to change the Dict, just the entries, `"core2", `"nehalem"` etc. So you know that the keys are arches, and not say, the life work of Shakespeare and you can reflect on the instances of the enum.

Reply:
If we change the keys like `"core2"` from a string to an enum type, how do we add a new value to that enum type at runtime?  What I'm considering is that Julia 1.6 is around for 3+ years, and in that time we want to add new ISAs, which we will be able to do from within a package by running something like `Base.BinaryPlatforms.ISAs_by_family["cannonlake"] = ...`.

Reply:
If we want to be able to interactively add new ISA's at runtime from packages then it clearly can't be an enum. I didn't know that was desireable, especially when it seems like there are multiple data structures here that need to be kept in sync and that the features are generated from the C-code. But if that's the case, then yeah, can't use an enum.

Reply:
> which we will be able to do from within a package by running something like `Base.BinaryPlatforms.ISAs_by_family["x86_64"]["cannonlake"] = ....`

Note that nothing from this PR will be a stable API.

Reply:
I changed how `arch_march_isa_mapping` is defined below, which should make it easier to keep that dictionary in sync with this one.